agents/systems/WIDGET-INTEGRATION-GUIDE.md

```diff
@@ -0,0 +1,678 @@
+# Widget Integration Developer Guide
+
+This guide explains how to create widgets that integrate with third-party services using the Castmill Widget Integration System.
+
+## Table of Contents
+
+1. [Overview](#overview)
+2. [Integration Types](#integration-types)
+3. [Credential Scopes](#credential-scopes)
+4. [Creating an Integration](#creating-an-integration)
+5. [PULL Mode Integration](#pull-mode-integration)
+6. [PUSH Mode Integration](#push-mode-integration)
+7. [Player Implementation](#player-implementation)
+8. [Testing](#testing)
+9. [Examples](#examples)
+
+## Overview
+
+The Widget Integration System provides a standardized way for widgets to connect with external services. It handles:
+
+- **Credential Management**: Secure storage and encryption of API keys and credentials
+- **Data Fetching**: Automated PULL of data from third-party APIs
+- **Webhook Handling**: PUSH of data from third-party services
+- **Version Control**: Efficient polling with version-based caching
+- **Error Handling**: Graceful degradation when services are unavailable
+
+## Integration Types
+
+### PULL Mode
+
+The backend periodically fetches data from a third-party API and caches it. Players poll for updates.
```

Comment:
Players (Devices) should not poll unless the player has been offline and after coming back online wants to check if there is new data, otherwise the backend will use web sockets to push the new data to all online players that care about this particular widget. I.e. players need to join the channel associated to a given widget instantiation (provided the player has the right to do so, for instance the player aka device is in the same organization as the instantiated widget.